torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -665,11 +669,27 @@ def view_avoid_dupes_with_primals(t):
         traced_tangents = pytree.tree_map(
             view_avoid_dupes_with_primals, traced_tangents
         )
-        # See Note [Tangents must be contiguous]
-        traced_tangents = pytree.tree_map(
-            coerce_tangent,
-            traced_tangents,
-        )
+
+        def recursive_suggest_memory_format(t):
+            suggest_memory_format = torch._prims_common.suggest_memory_format
+            if is_traceable_wrapper_subclass(t):
+                return [
+                    recursive_suggest_memory_format(getattr(t, attr))
```

Comment:
I'm a bit torn because:

(1) right now you are trying to track the memory formats of inner tensors separately from their outer subclasses. This technically isn't enough, because you need to handle nested subclasses (what if `getattr(t, attr)` is another tensor subclass that you need to recursively flatten and get the memory formats of its inner tensors?)

(2) you could fix that by having some recursive nested lists that contain the full hierarchy of suggested memory formats for every subclass tangent. I'm a bit worried about traversing those nested lists being bad for hot-path in the backward.

Alternatively, we could make the simplifying assumption that if a subclass advertises as "channels-last contiguous", it will likely also have channels-last-contiguous inner tensors (and vis versa), and therefore only bother to do this bookkeeping on the top-level subclass and not on the inner tensors.

Reply:
I guess it is totally possible to construct a subclass that violates that, though, and the recursive tracking shouldn't be *that* slow. So what do you think of:

(1) do the recursive tracking

(2) run a quick microbenchmark that times the backward of a function like this, and make sure its backward isn't a lot slower
```
def create_inp(x):
    return TwoTensor(TwoTensor(x.clone(), x.clone()), TwoTensor(x.clone(), x.clone()))

@torch.compile
def f(*args):
    return args

x = torch.randn(1, requires_grad=True)
inps = [create_inp(x) for _ in range(100)]
outs = f(inps)
# you probably just want to measure the overhead of the `CompiledFunction.backward()` directly, so you don't measure overhead from the rest of autograd
sum(outs).sum().backward()

```

Reply:
I think (maybe not in the first version) I did a recursive handling for subclasses attributes.
Will add test for NestedSubclasses to verify it.

Reply:
Benchmarked:
```
        benchmark_inps = [inps_fn(torch.randn(2, 3, 5, 5, requires_grad=True).to(memory_format=torch.channels_last)) for _ in range(100)]

        bwd_total_duration = 0
        for inps in benchmark_inps:
            outs = torch.compile(mc, backend="aot_eager", fullgraph=True)(*inps)
            s = outs[0].sum()
            time_start = time.time()
            s.backward()
            bwd_duration = time.time() - time_start
            bwd_total_duration += bwd_duration
        
        avg_bwd_duration = bwd_total_duration / len(benchmark_inps)
        print(f"XXX    SUBCLASS_GROUP avg_bwd_duration:{avg_bwd_duration*1000} ms")
        
        class M2(torch.nn.Module):
            def __init__(self) -> None:
                super().__init__()
                self.conv = torch.nn.Conv2d(3, 3, 3)

            def forward(self, x0, x1, x2, x3):
                return self.conv(x0), self.conv(x1), self.conv(x2), self.conv(x3)

        m2 = M2()
        m2.to(memory_format=torch.channels_last)
        m2.train()

        def inps_fn2(x):
            return (
                x.clone(),
                x.clone(),
                x.clone(),
                x.clone()
            )

        benchmark_inps2 = [inps_fn2(torch.randn(2, 3, 5, 5, requires_grad=True).to(memory_format=torch.channels_last)) for _ in range(100)]

        bwd_total_duration = 0
        for inps in benchmark_inps2:
            outs = torch.compile(m2, backend="aot_eager", fullgraph=True)(*inps)
            s = outs[0].sum()
            time_start = time.time()
            s.backward()
            bwd_duration = time.time() - time_start
            bwd_total_duration += bwd_duration
        
        avg_bwd_duration = bwd_total_duration / len(benchmark_inps)
        print(f"XXX NO_SUBCLASS_GROUP avg_bwd_duration:{avg_bwd_duration * 1000} ms")
```

XXX    SUBCLASS_GROUP avg_bwd_duration:7.817392349243163 ms
XXX NO_SUBCLASS_GROUP avg_bwd_duration:3.940465450286865 ms
Subclasses bwd is 2 times slower avg

Reply:
oh that delta between subclass and no delta is a good datapoint (that's also probably telling us that other things like the subclass `flatten/unflatten` we do in the subclass path adds up to quite a bit of overhead in general?)

My specific question though was more about if this PR causes any regressions for the existing subclass path. So "SUBCLASS_GROUP" (with your changes) vs. "SUBCLASS_GROUP" (without your changes)

Reply:
Measured:

Without changes:
channels_last: 11.92ms

With changes:
contiguous: 10.68ms
channels_last: 8ms

No regression with changes, used all-contiguous() to remove the difference of having contiguous() without-changes. But still do not understand fully why it speeds up the backward in case of all contiguous.